skr/webhook/internal/handler.go

```diff
@@ -0,0 +1,343 @@
+	}
+	if !validation.allowed {
+		h.Logger.Info(
+			fmt.Sprintf("%s %s %s", admissionReview.Request.Kind.Kind,
+				string(admissionReview.Request.Operation), InvalidationMessage),
+		)
+	} else {
+		h.Logger.Info(
+			fmt.Sprintf("%s %s %s", admissionReview.Request.Kind.Kind,
+				string(admissionReview.Request.Operation), ValidationMessage),
+		)
+	}
+
+	bytes, err := json.Marshal(&finalizedAdmissionReview)
+	if err != nil {
+		h.httpError(writer, http.StatusInternalServerError, fmt.Errorf("%s %w", AdmissionError, err))
+		return nil
+	}
+	return bytes
+}
+
+func (h *Handler) validateResources(writer http.ResponseWriter, admissionReview *admissionv1.AdmissionReview,
+	resourcesList []Resource, kymaName string,
+) validateResource {
+	if admissionReview.Request.Operation != admissionv1.Update {
+		return h.validAdmissionReviewObj(
+			fmt.Sprintf("%s operation not supported", admissionReview.Request.Operation), metav1.StatusFailure,
+		)
+	}
+	var message string
+	var status string
+	for _, resource := range resourcesList {
+		if admissionReview.Request.Kind.String() != resource.GroupVersionKind.String() {
+			continue
+		}
+		// Note: OldObject is nil for "CONNECT" and "CREATE" operations
+		// old object
+		oldObjectWatched := ObjectWatched{}
+		validatedResource := h.unmarshalRawObj(writer, admissionReview.Request.OldObject.Raw,
+			&oldObjectWatched, resource.GroupVersionKind.String())
+		if !validatedResource.allowed {
+			return validatedResource
+		}
+
+		// new object
+		objectWatched := ObjectWatched{}
+		validatedResource = h.unmarshalRawObj(writer, admissionReview.Request.Object.Raw,
+			&objectWatched, resource.GroupVersionKind.String())
+		if !validatedResource.allowed {
+			return validatedResource
+		}
+
+		if resource.Status {
+			if !reflect.DeepEqual(oldObjectWatched.Status, objectWatched.Status) {
+				message = "sent requests to KCP for Status"
+			}
+		}
+
+		if resource.Spec && message == "" {
+			if !reflect.DeepEqual(oldObjectWatched.Spec, objectWatched.Spec) {
+				message = "sent requests to KCP for Spec"
+			}
+		}
+
+		if message != "" {
+			status = h.sendRequestToKcp(kymaName, objectWatched)
+		}
+
+		// since resource was found - exit loop
+		break
+	}
+
+	// send valid admission response - under all circumstances!
+	return h.validAdmissionReviewObj(message, status)
+}
+
+func (h *Handler) sendRequestToKcp(kymaName string, watched ObjectWatched) string {
+	// send request to kcp
+	watcherEvent := &types.WatcherEvent{
+		KymaCr:    kymaName,
+		Namespace: watched.Namespace,
+		Name:      watched.Name,
+	}
+	postBody, err := json.Marshal(watcherEvent)
+	if err != nil {
+		h.Logger.Error(err, "")
+		return metav1.StatusFailure
+	}
+
+	responseBody := bytes.NewBuffer(postBody)
+
+	kcpIP := os.Getenv("KCP_IP")
+	kcpPort := os.Getenv("KCP_PORT")
+	contract := os.Getenv("KCP_CONTRACT")
+	component := "manifest"
+
+	if kcpIP == "" || kcpPort == "" || contract == "" {
+		return metav1.StatusSuccess
```

Comment:
why do you return success here, but failure on the marshal? are we not supposed to be completely transparent and always return success on failing anything?

Reply:
agree, will change everything to success